litestar/utils/predicates.py

```diff
@@ -124,6 +124,9 @@ def is_dataclass_class(annotation: Any) -> TypeGuard[type[DataclassProtocol]]:
         ``True`` if instance or type of ``dataclass``.
     """
     try:
+        origin = get_origin_or_inner_type(annotation)
```

Comment:
This is just for removing wrapper types like `Annotated` et al. isn't it? I.e., if `annotation` is `Annotated[list[int]]` then this will return `list[int]`? When I think about annotations and inner types, I think of `int` in `list[int]` being the inner type.

Not your issue, but do you think the name `get_origin_or_inner_type()` is a big confusing? `unwrap_and_get_origin()` would maybe be a better name. WDYT?

Reply:
I definitely agree. I had to go read the code to make sure it really is only giving me back the origin class :P

Should this be included in this PR itself or a separate one? Also, is this public API? It's not there in the documentation, but it's also not under a private namespace of `_utils`. I'm leaning towards it's not public since it's not present in the docs.

Reply:
> Should this be included in this PR itself or a separate one?

I think better ti do it separately  and keep this PR focussed, i was just curious what your thoughts on it were👍

> Also, is this public API? It's not there in the documentation, but it's also not under a private namespace of `_utils`. I'm leaning towards it's not public since it's not present in the docs.

I think we can rename the function that we use internally and alias it to the current name, with a deprecation for 3.0.

Reply:
Yeah this sounds like a good plan. On that note, what do you think about moving some of the pydantic predicates etc into a `litestar.contrib.pydantic.predicates` module?

Reply:
It would be nice if all of the pydantic stuff could exist inside the contrib/plugin eventually, but I don't think we should be importing from contrib into our core. So, while we are using funcs like `is_pydantic_constrained_field` and `is_pydantic_model_instance` outside of the contrib package, then we really need to keep them in there.

If we can get it to the point where our core doesn't have to care at all about third party libraries, it means we have the abstraction right and no coupling between our core and pydantic, or attrs or whatever. :pinched_fingers:

Reply:
The more I look at this, even `unwrap_and_get_origin()` wouldn't be appropriate, as this implies that it behaves the same as `get_origin()` while ignoring "wrapper" types, but it doesn't because it returns the value from the `instantiable_type_mapping`. So it would have to be more like `get_unwrapped_instantiable_origin()` or something.

Reply:
Well, this doesn't necessarily return an instantiable origin. It defaults to the origin if it's not part of the instantiable origin map. So, then the name would have to be `unwrap_and_get_instantiable_origin_if_it_exists_else_the_origin` 😆

Reply:
I created issue #2467 so we can track this.